File: app/src/main/java/org/openobservatory/ooniprobe/common/worker/UpdateDescriptorsWorker.java
package org.openobservatory.ooniprobe.common.worker;

import android.Manifest;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.openobservatory.ooniprobe.R;
import org.openobservatory.ooniprobe.common.Application;
import org.openobservatory.ooniprobe.common.ThirdPartyServices;
import org.openobservatory.ooniprobe.domain.TestDescriptorManager;
import org.openobservatory.ooniprobe.model.database.TestDescriptor;

import java.util.ArrayList;

public class UpdateDescriptorsWorker extends Worker {
    public static final String UPDATED_DESCRIPTORS_WORK_NAME = String.format("%s.UPDATED_DESCRIPTORS_WORK_NAME", UpdateDescriptorsWorker.class.getName());
    private static final String TAG = UpdateDescriptorsWorker.class.getSimpleName();
    public static final String UPDATE_DESCRIPTOR_CHANNEL = UpdateDescriptorsWorker.class.getSimpleName();
    private static final String KEY_UPDATED_DESCRIPTORS = String.format("%s.KEY_UPDATED_DESCRIPTORS", UpdateDescriptorsWorker.class.getName());

    public UpdateDescriptorsWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {

        try {
            Context applicationContext = getApplicationContext();

            Log.d(TAG, "Fetching descriptors from input");
            makeStatusNotification("Starting update for descriptors");

            ArrayList<TestDescriptor> updatedDescriptors = new ArrayList<>();

            for (TestDescriptor descriptor : TestDescriptorManager.descriptorsWithAutoUpdateEnabled()) {

                Log.d(TAG, String.format("Fetching updates for %d ", descriptor.getRunId()));
                makeStatusNotification(String.format("Fetching updates for %s ", descriptor.getName()));

                TestDescriptor updatedDescriptor = TestDescriptorManager.fetchDescriptorFromRunId(descriptor.getRunId(), applicationContext);

                if (updatedDescriptor.getVersion() > descriptor.getVersion()) {

                    updatedDescriptor.setAutoUpdate(descriptor.isAutoUpdate());
                    updatedDescriptor.setAutoRun(descriptor.isAutoRun());

                    Log.d(TAG, String.format("Saving updates for %d ", descriptor.getRunId()));
                    makeStatusNotification(String.format("Saving updates for %s", descriptor.getName()));

                    updatedDescriptor.save();
                    updatedDescriptors.add(updatedDescriptor);
                }

            }

            Data outputData = new Data.Builder().putString(KEY_UPDATED_DESCRIPTORS, ((Application) applicationContext).getGson().toJson(updatedDescriptors)).build();
            makeStatusNotification("Descriptor updates complete");
            return Result.success(outputData);

        } catch (Exception exception) {
            Log.e(TAG, "Error Updating");
            makeStatusNotification("Error Updating");
            exception.printStackTrace();
            ThirdPartyServices.logException(exception);
            return Result.failure();
        }
    }

    /**
     * Create a Notification that is shown as a heads-up notification if possible.
     *
     * @param message Message shown on the notification
     */
    private void makeStatusNotification(String message) {

        // Make a channel if necessary
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create the NotificationChannel, but only on API 26+ because
            // the NotificationChannel class is new and not in the support library
            int importance = NotificationManager.IMPORTANCE_HIGH;
            NotificationChannel channel =
                    new NotificationChannel(UPDATE_DESCRIPTOR_CHANNEL, "Run Descriptor Updates", importance);
            channel.setDescription("Shows notification related to updates being made to Run Descriptors");

            // Add the channel
            NotificationManager notificationManager =
                    (NotificationManager) getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);

            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
            }
        }

        // Create the notification
        NotificationCompat.Builder builder = new NotificationCompat.Builder(getApplicationContext(), UPDATE_DESCRIPTOR_CHANNEL)
                .setSmallIcon(R.drawable.notification_icon)
                .setContentTitle("Descriptor Update")
                .setContentText(message)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setVibrate(new long[0]);

        // Show the notification
        if (ActivityCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        NotificationManagerCompat.from(getApplicationContext()).notify((int) System.currentTimeMillis(), builder.build());
    }
}

File: app/src/main/java/org/openobservatory/ooniprobe/common/worker/UpdateDescriptorsWorker.kt
package org.openobservatory.ooniprobe.common.worker

import android.content.Context
import android.util.Log
import androidx.work.Data
import androidx.work.Worker
import androidx.work.WorkerParameters
import org.openobservatory.ooniprobe.common.Application
import org.openobservatory.ooniprobe.common.PreferenceManager
import org.openobservatory.ooniprobe.common.TestDescriptorManager
import org.openobservatory.ooniprobe.common.ThirdPartyServices
import org.openobservatory.ooniprobe.model.database.TestDescriptor
import org.openobservatory.ooniprobe.model.database.shouldUpdate
import javax.inject.Inject

var d: UpdateDescriptorsWorkerDependencies = UpdateDescriptorsWorkerDependencies()
const val PROGRESS = "PROGRESS"

class AutoUpdateDescriptorsWorker(
        context: Context,
        workerParams: WorkerParameters
) : Worker(context, workerParams) {

    override fun doWork(): Result {

        setProgressAsync(Data.Builder().putInt(PROGRESS, 0).build())

        val app = applicationContext.applicationContext as Application
        app.serviceComponent.inject(d)

        return try {
            Log.d(TAG, "Fetching descriptors from input")

            val updatedDescriptors: ArrayList<TestDescriptor> = ArrayList()

            for (descriptor in d.testDescriptorManager.getRunV2Descriptors(
                    autoUpdate = true,
                    expired = false
            )) {

                Log.d(TAG, "Fetching updates for ${descriptor.runId}")

                d.testDescriptorManager.fetchDescriptorFromRunId(
                        descriptor.runId,
                        applicationContext
                )?.let { updatedDescriptor ->
                    if (descriptor.shouldUpdate(updatedDescriptor)) {
                        updatedDescriptor.isAutoUpdate = descriptor.isAutoUpdate

                        Log.d(TAG, "Saving updates for ${descriptor.runId}")

                        updatedDescriptor.save()
                        updatedDescriptors.add(updatedDescriptor)
                    }
                }
            }

            val outputData = Data.Builder()
                .putString(
                    KEY_UPDATED_DESCRIPTORS,
                    (applicationContext as Application).gson.toJson(updatedDescriptors)
                ).build()

            Log.e(TAG, "Descriptor updates complete")

            Result.success(outputData)

        } catch (exception: Exception) {
            Log.e(TAG, "Error Updating")
            exception.printStackTrace()
            ThirdPartyServices.logException(exception)
            Result.failure()
        }
    }

    companion object {
        @JvmField
        var UPDATED_DESCRIPTORS_WORK_NAME =
                "${AutoUpdateDescriptorsWorker::class.java.name}.UPDATED_DESCRIPTORS_WORK_NAME"

        private val TAG = AutoUpdateDescriptorsWorker::class.java.simpleName

        private val UPDATE_DESCRIPTOR_CHANNEL: String =
                AutoUpdateDescriptorsWorker::class.java.simpleName

        private val KEY_UPDATED_DESCRIPTORS =
                "${AutoUpdateDescriptorsWorker::class.java.name}.KEY_UPDATED_DESCRIPTORS"
    }
}

class ManualUpdateDescriptorsWorker(
        context: Context,
        workerParams: WorkerParameters
) : Worker(context, workerParams) {

    override fun doWork(): Result {

        setProgressAsync(Data.Builder().putInt(PROGRESS, 0).build())

        val app = applicationContext.applicationContext as Application
        app.serviceComponent.inject(d)

        return try {
            Log.d(TAG, "Fetching descriptors from input")

            val updatedDescriptors: ArrayList<TestDescriptor> = ArrayList()

            val descriptors = inputData.getLongArray(KEY_DESCRIPTOR_IDS)?.let {
                d.testDescriptorManager.getDescriptorsFromIds(it.toTypedArray())
            } ?: run {
                d.testDescriptorManager.getRunV2Descriptors(expired = false)
            }

            if (descriptors.isEmpty()) {
                Log.e(TAG, "No descriptors to update")
                return Result.success()
            }

            for (descriptor in descriptors) {
                Log.d(TAG, "Fetching updates for ${descriptor.runId}")

                d.testDescriptorManager.fetchDescriptorFromRunId(
                        descriptor.runId,
                        applicationContext
                )?.let { updatedDescriptor ->

                    /**
                     * NOTE(aanorbel): Refine this logic to only update if the descriptor has changed.
                     * Consider explicit version compare.
                     */
                    if (descriptor.shouldUpdate(updatedDescriptor)) {
                        if (descriptor.isAutoUpdate) {
                            updatedDescriptor.isAutoUpdate = true

                            Log.d(TAG, "Saving updates for ${descriptor.runId}")

                            updatedDescriptor.save()
                        } else {
                            Log.d(TAG, "Not saving updates for ${descriptor.runId}")
                            updatedDescriptors.add(updatedDescriptor)
                        }
                    }
                }
            }
            val outputData = Data.Builder()
                .putString(
                    KEY_UPDATED_DESCRIPTORS,
                    (applicationContext as Application).gson.toJson(updatedDescriptors)
                ).build()

            Log.e(TAG, "fetching updates complete")

            setProgressAsync(Data.Builder().putInt(PROGRESS, 100).build())
            if (updatedDescriptors.isEmpty()) {
                Log.e(TAG, "No descriptors were updated")
                Result.success()
            } else {
                Result.success(outputData)
            }

        } catch (exception: Exception) {
            Log.e(TAG, "Error Updating")
            exception.printStackTrace()
            ThirdPartyServices.logException(exception)
            Result.failure()
        }
    }

    companion object {
        @JvmField
        var UPDATED_DESCRIPTORS_WORK_NAME =
                "${ManualUpdateDescriptorsWorker::class.java.name}.UPDATED_DESCRIPTORS_WORK_NAME"

        private val TAG = ManualUpdateDescriptorsWorker::class.java.simpleName

        private val UPDATE_DESCRIPTOR_CHANNEL: String =
                ManualUpdateDescriptorsWorker::class.java.simpleName

        @JvmField
        var KEY_UPDATED_DESCRIPTORS =
                "${ManualUpdateDescriptorsWorker::class.java.name}.KEY_UPDATED_DESCRIPTORS"

        @JvmField
        var KEY_DESCRIPTOR_IDS =
                "${ManualUpdateDescriptorsWorker::class.java.name}.KEY_DESCRIPTOR_IDS"
    }
}

class UpdateDescriptorsWorkerDependencies {
    @Inject
    lateinit var testDescriptorManager: TestDescriptorManager

    @Inject
    lateinit var preferenceManager: PreferenceManager
}
